### ISSUE TITLE
Cuda support for matrix multiplication on Long Tensor

### ISSUE THREAD

**Issue by LaRiffle** [pytorch/pytorch#44428]
## 🚀 Feature
I would love cuda support for matrix multiplication on Long Tensor
Currently:
```
import torch

x = torch.LongTensor([[1]])
x = x.cuda()
x + x  # works
x * x  # works
x @ x  # FAILS
```

## Motivation & Pitch

I'm working on privacy preserving machine learning which uses a lot LongTensors (like the CrypTen team cc @jspisak ), and having matmul accelerated with cuda (like it is for normal matmul on floats) would be game changing.

Is there any plan to support this, or any work around?

Thanks a lot for all the efforts.

**Comment by gchanan** [comment]
doesn't have to do with dispatch, which refers to a different type of dispatching (i.e. at the DispatchKey level, not the Dispatch macro level).

**Comment by ezyang** [comment]
cc @pbelevich (but he's on parental leave right now so don't expect a response any time soon)

**Comment by ezyang** [comment]
cublas doesn't support integer types (https://docs.nvidia.com/cuda/cublas/index.html#cublas-level-1-function-reference) but we could probably provide a "shitty" matrix multiply implementation on integers without too much work; I believe this is what @pbelevich did on CPU for crypten

**Comment by jspisak** [comment]
Hi - sorry i missed this ping. Yes, we have this on our radar but it isn't resourced for this half. Is there a prioritized set of Ops that might work for a PoC?

**Comment by lvdmaaten** [comment]
In CrypTen, we currently work around these issues by doing the actual computations on DoubleTensors with some magic that encodes/decodes between the long and double types. This may be interesting for @LaRiffle team as well, if only as a temporary work-around: https://github.com/facebookresearch/CrypTen/blob/master/crypten/cuda/cuda_tensor.py

**Comment by LaRiffle** [comment]
Thanks a lot for pointing to this work-around! That's very interesting, I'll investigate in that direction 🙂

**Comment by dizcza** [comment]
I'd love to have addmm_cuda for integers (both int32 and int64) as well!

I'm working in neuroscience and integer computations happen often in this field. I'm wondering how [BindsNET](https://github.com/BindsNET/bindsnet) project alleviates this limitation. Perhaps, they also convert everything to floats. However, in my case, it'd be cumbersome to write the logic in floats explicitly.

Example: Binomial and Bernoulli distributed input data and weights, winner-take-all algorithms - they all work with integers.

**Comment by ezyang** [comment]
It doesn't seem like it would be too much work to steal crypten's trick for reusing the existing float kernels; writing a crappy CUDA matrix multiply also shouldn't be too much work either.

**Comment by ngimel** [comment]
On consumer cards, double precision gemm performance will be catastrophically slow.

**Comment by ailzhang** [comment]
@LaRiffle How much do you care about perf for these kernels?

**Comment by v0dro** [comment]
I'd like to take up this issue. To summarize, here's what I understand so far:

* Integer matrix multiplication using CUDA is slow since it makes use of CUBLAS kernels, which are written for floating point math.
* its not clear whether a hand-written matmul implementation will be faster or conversion to floating point + calling cublas.
* We need to understand this first and then implement the matmul kernel for CUDA.

Let me try to come up both implementations, benchmark them, and see which one is faster for some sample problem sizes. The cublas conversion should be pretty straightforward to implement IMO.

Assigning this to myself now.

**Comment by Lezcano** [comment]
@LaRiffle @dizcza Is loss of precision also an issue here, or is it mostly about convenience + memory footprint of having to call the casts?

**Comment by LaRiffle** [comment]
Hey @Lezcano & @v0dro! Loss of precision is an issue indeed: one big usage is for cryptography applications (including the ones from Crypten), for which we need exact computation.

**Comment by dizcza** [comment]
For me, it's rather about the inconvenience of casting between int and float back and forth.
I finally switched everything to floats. Although it's not clear why I did this for a non-developer, it's fast.

**Comment by v0dro** [comment]
Removing my assignment to this issue since the PR is ready and a consensus whether to merge it or not is yet to be reached.
